alibi_detect/cd/tabular.py

```diff
@@ -75,16 +80,19 @@ def __init__(
         self.x_ref_categories, self.cat_vars = {}, []  # no categorical features assumed present
         if isinstance(categories_per_feature, dict):
             vals = list(categories_per_feature.values())
+            int_types = (int, np.int16, np.int32, np.int64)
             if all(v is None for v in vals):  # categories_per_feature = Dict[int, NoneType]
                 x_flat = self.x_ref.reshape(self.x_ref.shape[0], -1)
                 categories_per_feature = {f: list(np.unique(x_flat[:, f]))  # type: ignore
                                           for f in categories_per_feature.keys()}
-            elif all(isinstance(v, (int, np.int16, np.int32, np.int64)) for v in vals):
+            elif all(isinstance(v, int_types) for v in vals):
                 # categories_per_feature = Dict[int, int]
                 categories_per_feature = {f: list(np.arange(v))  # type: ignore
                                           for f, v in categories_per_feature.items()}
-            elif not all(isinstance(v, list) for v in vals):
-                raise NotImplementedError  # categories_per_feature not Dict[int, list]
+            elif not all(isinstance(v, list) for v in vals) and \
+                    all(isinstance(v, int_types) for val in vals for v in val):  # type: ignore
+                raise NotImplementedError('categories_per_feature needs to be None or one of '
```

Comment:
I don't think the default `None` case is actually handled (unlike in chisquare.py)?

Reply:
Oh I think see why, for `TabularDrift` the `None` case means "no categorical features" so no handling required, whereas for `ChiSquareDrift` the `None` case is more meaningful (infer all categories).

Reply:
Yes, it should be clear in the docstrings / documentation.